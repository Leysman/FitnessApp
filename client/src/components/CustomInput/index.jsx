// CustomInput.jsx
import { useState } from 'react';
import { useField, useFormikContext } from 'formik';
import { FiEye, FiEyeOff } from 'react-icons/fi';
import classNames from 'classnames';
import styles from './CustomInput.module.scss';

const CustomInput = ({ id, label, type = 'text', ...props }) => {
  const [field, meta] = useField(props);
  const { submitCount } = useFormikContext();
  const showError = Boolean(meta.error) && (meta.touched || submitCount > 0);

  const [visible, setVisible] = useState(false);
  const isPassword = type === 'password';
  const inputType = isPassword && visible ? 'text' : type;

  return (
    <div className={classNames(
      styles.inputWrapper,
      showError && styles.error
    )}>
      {label && <label htmlFor={id}>{label}</label>}
      <div className={styles.fieldInner}>
        <input
          id={id}
          {...field}
          {...props}
          type={inputType}
          aria-invalid={showError}
        />
        {isPassword && (
          <button
            type="button"
            className={styles.toggle}
            onClick={() => setVisible(v => !v)}
            aria-label={visible ? 'Hide password' : 'Show password'}
          >
            {visible ? <FiEyeOff /> : <FiEye />}
          </button>
        )}
      </div>
      {showError && (
        <p className={styles.errorMessage}>{meta.error}</p>
      )}
    </div>
  );
};

export default CustomInput;
