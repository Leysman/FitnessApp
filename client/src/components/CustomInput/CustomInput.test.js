import { render } from '@testing-library/react';
import { Formik } from 'formik';
import CustomInput from '.';

describe('CustomInput snapshots', () => {

  test('renders regular input (no format)', () => {
    const { asFragment } = render(
      <Formik initialValues={{ name: '' }} onSubmit={() => {}}>
        <CustomInput id="name" name="name" label="Name" />
      </Formik>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders formatted input (with format)', () => {
    const { asFragment } = render(
      <Formik initialValues={{ phone: '' }} onSubmit={() => {}}>
        <CustomInput id="phone" name="phone" label="Phone" format="+38 (###) ###-##-##" />
      </Formik>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders input with error', () => {
    const { asFragment } = render(
      <Formik
        initialValues={{ email: '' }}
        initialErrors={{ email: 'Email is required' }}
        initialTouched={{ email: true }}
        onSubmit={() => {}}
      >
        <CustomInput id="email" name="email" label="Email" />
      </Formik>
    );
    expect(asFragment()).toMatchSnapshot();
  });

});
