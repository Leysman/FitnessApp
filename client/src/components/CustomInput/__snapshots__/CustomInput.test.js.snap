// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CustomInput snapshots renders formatted input (with format) 1`] = `
<DocumentFragment>
  <div
    class="inputWrapper"
  >
    <label
      for="phone"
    >
      Phone
    </label>
    <div
      class="fieldInner"
    >
      <input
        aria-invalid="false"
        format="+38 (###) ###-##-##"
        id="phone"
        name="phone"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`CustomInput snapshots renders input with error 1`] = `
<DocumentFragment>
  <div
    class="inputWrapper error"
  >
    <label
      for="email"
    >
      Email
    </label>
    <div
      class="fieldInner"
    >
      <input
        aria-invalid="true"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
    <p
      class="errorMessage"
    >
      Email is required
    </p>
  </div>
</DocumentFragment>
`;

exports[`CustomInput snapshots renders regular input (no format) 1`] = `
<DocumentFragment>
  <div
    class="inputWrapper"
  >
    <label
      for="name"
    >
      Name
    </label>
    <div
      class="fieldInner"
    >
      <input
        aria-invalid="false"
        id="name"
        name="name"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
